lib/swaps/Swaps.ts

```diff
@@ -243,16 +243,21 @@ class Swaps extends EventEmitter {
    * A promise wrapper for a swap procedure
    * @param maker the remote maker order we are filling
    * @param taker our local taker order
-   * @returns A promise that is resolved once the swap is completed, or rejects otherwise
+   * @returns A promise that resolves to a [[SwapSuccess]] once the swap is completed, or a [[SwapFailureReason]] if it fails
    */
-  public executeSwap = async (maker: PeerOrder, taker: OwnOrder): Promise<SwapSuccess> => {
+  public executeSwap = async (maker: PeerOrder, taker: OwnOrder): Promise<SwapSuccess | SwapFailureReason> => {
     await this.verifyExecution(maker, taker);
-    const rHash = await this.beginSwap(maker, taker);
-    if (!rHash) {
-      throw new Error('cannot execute swap. rHash not found');
+    const beginResult = await this.beginSwap(maker, taker);
+    let rHash: string;
+    if (typeof beginResult === 'string') {
```

Comment:
Optional: create another type called `SwapBeginSuccess` & `isSwapBeginSuccess` method, to make it easier to follow and understand why it's differed from `SwapSuccess`.

Reply:
The type of `SwapBeginSuccess` would just be a string right? That might be a little confusing as well as `isSwapBeginSuccess` would return true for any string. How about either:

1. I create an `isRHash` method that checks if the supplied parameter is a hex string of the correct length and use this here for line 252.
2. I create an `isSwapFailure` type check method and check for the failure first, `else` set the rHash value.

Reply:
Both are fine, but I think in this case rejecting the promise on failure will be the simplest.

Reply:
OK I'm going to refactor this by rejecting/throwing the failure reason and declaring the type in the `catch` blocks.